core/src/main/java/org/apache/iceberg/BaseUpdatePartitionSpec.java

```diff
@@ -133,7 +133,9 @@ public BaseUpdatePartitionSpec addField(String name, Term term) {
     Pair<Integer, String> validationKey = Pair.of(sourceTransform.first(), sourceTransform.second().toString());
 
     PartitionField existing = transformToField.get(validationKey);
-    Preconditions.checkArgument(existing == null,
+    Preconditions.checkArgument(existing == null ||
+        (deletes.contains(existing.fieldId()) &&
+            !existing.transform().dedupName().equals(sourceTransform.second().dedupName())),
```

Comment:
`dedupName` isn't correct. It is okay to drop a `months` field and replace it with a `days` field. But the dedup name for both is `time`. Instead, this should check the `toString` of the transform to ensure it isn't an identical transform.

Reply:
ah, sorry, my bad, updated